refactor(theme): move theme token variables to sass map

mattrosno opened the pull request:
- so we can use tokens in default and overrides mixin
- so we can override default theme by setting one global sass variable

Closes #395

#### Changelog

**New**

- `functions.scss` and `variables.scss` to support mixin changes
- `carbon--theme-overrides` mixin

**Changed**

- Previous theme mixins replaced with `carbon--theme` mixin that can pass in an optional theme name as parameter (`white`, `g10`, `g90`, or `g100`)

**Migration**

- `carbon-components` needs to change `@include carbon--theme-white()` to `@include carbon--theme`, and also needs to document how `$carbon--theme` global variable can be set to use a non-white theme

asudoh commented:
Thank you for the work @mattrosno! Here are my suggestions, so we don't have to lock this feature down to `carbon-elements`' `1.0` release:

* Keep mixins like `@mixin carbon--theme-g10`, by making it an alias to: `@include carbon--theme('g10')`
* Avoid changing the import path, or keep the old import path by making it an alias to the new import path

But if @joshblack agrees, just going with this PR is fine as the change seems fairly containable.

asudoh commented on packages/themes/tasks/build.js:
```diff
@@ -8,93 +8,139 @@
+  const variables = `${FILE_BANNER}
+${themeMaps}
+
+${initialVariables}`;
```
Wondering if we can separate this portion to a separate `.scss` file or surround this part by a mix-in, in order to avoid [`@import 'vendor/@carbon/elements/scss/themes/mixins'`](https://github.com/IBM/carbon-components/blob/v9.83.0/src/globals/scss/_theme-tokens.scss#L13) defining theme variables. Thanks!

joshblack replied:
@asudoh unfortunately for `carbon--themes` we need a reference to the default theme (white) in order to properly reset the theme values 😞

joshblack replied:
Actually, i'm not sure. The problem is that `carbon--theme` requires knowledge about the default theme in order to work, which means that themes needs variables already initialized.

asudoh replied:
Good point, I think the concern is about `$carbon--theme--white` reference in `@mixin carbon--theme($theme: $carbon--theme--white)` is it correct? From my PoV `$carbon--theme--*` including `$carbon--theme--white` can be left in what `_mixins.scss` imports. My concern is rather about each theme tokens defined as glonal, e.g. `$ui-01` - Wanted to make sure loading `_mixins.scss` won't automatically define `$ui-01`, etc.

joshblack commented on packages/themes/tasks/build.js:
```diff
@@ -8,93 +8,139 @@
+  // $carbon--theme--name: (
+  //   token-name: token-value
+  // ) !default !global;
+  // $theme--name: $carbon--theme--name;
```
Honestly just as a shorthand for convenience, no other reason 👍

mattrosno replied:
If we're using `carbon--` prefix for other "public" Sass variables/mixins/functions coming from elements, might as well encourage good habits and not provide the non-namespaced aliases?

joshblack commented:
@mattrosno good point on the default theme, I think there should be a `$carbon--theme` variable that we can then use as the default value in the `carbon--theme` mixin 👍

mattrosno commented on packages/themes/scss/_variables.scss:
```diff
@@ -0,0 +1,285 @@
+) !default !global;
+$theme--g100: $carbon--theme--g100;
+
+$interactive-01: map-get($carbon--theme, interactive-01) !default !global;
```
The variables were previously set in a mixin. In `carbon-components`, the previous mixin was called inside of a feature flag mixin. With these variables not set from inside a mixin, how would one conditionally use these variables? With them not in a mixin, the `@import` is setting the variables from carbon-elements regardless if the `components-x` feature flag is set or not.

```scss
@import './vendor/@carbon/elements/scss/themes/themes';

@mixin theme--experimental() {
  // Set theme variables here
}
```

mattrosno replied:
Related to not defining variables from within a mixin, if you wanted to set an alternate default theme, using `$carbon--theme` you would:

```scss
@import '../../src/globals/scss/vendor/@carbon/elements/scss/themes/variables';

$carbon--theme: $carbon--theme--g90;

@import '../../src/globals/scss/styles';
```

First you'd have to import the theme variables, so then you could globally set `$carbon--theme`. This original PR had the global override be a string like `$carbon--theme: 'g90';` so no preliminary imports were needed. Think we need to refactor to prevent the preliminary import?

This snippet **only** works if the variables are defined in a mixin, and that mixin is called **once** in `_theme-tokens.scss` since the variables are `!default`ed.

mattrosno commented on packages/themes/scss/_mixins.scss:
```diff
@@ -0,0 +1,74 @@
+//
+
+@import './theme';
+@import './variables';
```
I don't think `@import './variables';` is necessary as _mixins.scss does not reference anything from _variables.scss.

mattrosno commented:
@joshblack my recent commit moves variable declarations inside the mixin so importing doesn't automatically initialize the variables - so we can determine when we initialize variables to play nicely with carbon-components feature flag.

I also added a parameter to the mixin to conditionally set the `!defaults` keyword. Either we have two mixins, or one as I did. By having one, `$carbon--theme`, with naming not specific to "default" or "overrides", we can be forward thinking so if future refactoring means we can eliminate `!default` and `!global` flags over time, we can so in a non-breaking way. (May have to keep around a second param but not use it to prevent the breaking change.)

I also removed `theme--` aliases so we encourage proper use of namespacing. I have not documented yet; will do after our Sass output feels good.

With my change, in carbon-components you import the `themes.scss` and then in the feature flag mixin, call `@include carbon--theme();`.

Application usage:

```scss
// import theme variables only if need them to set a non-white default theme
@import '../../src/globals/scss/vendor/@carbon/elements/scss/themes/variables';

// uncomment both to apply the g90 theme with $interactive-01 being pink
// $carbon--theme: $carbon--theme--g90; // use g90 for default theme
// $interactive-01: pink; // or, instead, only globally define specific variables

// carbon components
@import '../../src/globals/scss/styles';

// second parameter omits `!default` in variable reassignments
@include carbon--theme($carbon--theme--g90, false) {
  .my-dark-theme .foo {
    color: $interactive-03;
  }
}
```

joshblack commented:
> my recent commit moves variable declarations inside the mixin so importing doesn't automatically initialize the variables - so we can determine when we initialize variables to play nicely with carbon-components feature flag.

Is there any harm to the auto-init? Will solve this problem in future sass versions and they should all have `!default` so they will play nicely with carbon-components
